src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -197,6 +206,7 @@ internal void OnInitialSettingsAckReceived(Http2Connection connection)
             internal void OnDataOrHeadersReceived(Http2Connection connection)
             {
                 if (_state != State.Waiting) return;
+                if (_sendPolicyCounter >= MaxPingsWithoutSending) return;
```

Comment:
Is this supposed to be a hard limit?
Am I reading the code correctly? This will check the value before putting the PING into the write queue, but there might be some unprocessed pending PING frames which would increment the counter, thus leading to more than `MaxPingsWithoutSending` PINGs sent. Or I'm missing something here?

Reply:
You are right, I will try to do something in `SendPingAsync` about this.

Reply:
Thinking about this, it looks like there is no easy way to get rid of these races without introducing a lot of additional complexity around this code, and I'm not sure if it's worth it.

I'm tempted to close this PR and the issue #62216, since I'm not sure we are trying to solve a problem that actually exist in practice after google's backend tuning. We can come back and work on a more sophisticated solution if we get new user reports.